### PR TITLE
Read provider version from .terraform.lock.hcl lock file when possible

### PR DESCRIPTION
**Pull request by sundowndev**
| Q                 | A
| ----------------- | ---
| 🐛 Bug fix?       | no
| 🚀 New feature?   | yes
| ⚠ Deprecations?   | no
| ❌ BC Break       | no
| 🔗 Related issues | #929
| ❓ Documentation  | https://github.com/cloudskiff/driftctl-docs/pull/130 

## Description

This PR implement a lock package that parses the content of a Terraform lock file. This logic is used in the Terraform supplier in order to install a specific version when we found a Terraform lock file in at the given path.

### CODE REVIEW

File: pkg/remote/aws/provider.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package aws
 
 import (
 	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/hashicorp/terraform/addrs"
```
Comment (eliecharra) [inline pkg/remote/aws/provider.go:5]: We cannot use this struct anymore since in the newer version this is unexported (moved under `internal` pkg)

We should have our own struct instead under `pkg/terraform`

File: pkg/remote/terraform/lockfile.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+	"github.com/hashicorp/hcl/v2/gohcl"
+	"github.com/hashicorp/hcl/v2/hclparse"
+	"github.com/hashicorp/terraform/addrs"
+	"github.com/hashicorp/terraform/tfdiags"
+)
+
+type LockFileProvider struct {
+	Addr        addrs.Provider
+	Constraints string   `hcl:"constraints"`
+	Hashes      []string `hcl:"hashes"`
```
Comment (eliecharra) [inline pkg/remote/terraform/lockfile.go:17]: Can we remove theses fields from this struct ? We only want the version.

File: pkg/remote/terraform/provider.go [hunk 3]
```diff
@@ -24,13 +25,16 @@ import (
 	tf "github.com/cloudskiff/driftctl/pkg/terraform"
 )
 
+const LockFileName = ".terraform.lock.hcl"
```
Comment (eliecharra) [inline pkg/remote/terraform/provider.go:28]: I don't think we need this exported const since this is only used at 1 place below.

File: pkg/remote/terraform/lockfile.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
+package terraform
```
Comment (eliecharra) [inline pkg/remote/terraform/lockfile.go:1]: It could be nice to put this in `pkg/terraform/lock` instead of remote, this is not really related to driftctl remote stuff.

We should probably only expose a single method from this pkg to return the related provider config from this pkg.

We may return an error (or terraform diags) too and log it in debug

File: pkg/remote/terraform/provider.go [hunk 5]
```diff
@@ -88,8 +92,16 @@ func (p *TerraformProvider) Runner() *parallel.ParallelRunner {
 }
 
 func (p *TerraformProvider) configure(alias string) error {
+	var version string
+	lockFile, diags := LoadLocksFromFile(LockFileName)
```
Comment (eliecharra) [inline pkg/remote/terraform/provider.go:96]: We should log in debug the diags, it will help us investigate some case where we wont be able to read the lockfile for some reason

**Review by eliecharra — CHANGES_REQUESTED** [review]
I left some architectural comments, I will do another deeper review pass after that

File: pkg/remote/terraform/provider.go [hunk 6]
```diff
@@ -88,8 +89,14 @@ func (p *TerraformProvider) Runner() *parallel.ParallelRunner {
 }
 
 func (p *TerraformProvider) configure(alias string) error {
+	var version string
+	lockFile, _ := lock.LoadLocksFromFile(".terraform.lock.hcl")
```
Comment (eliecharra) [inline pkg/remote/terraform/provider.go:93]: My first comment is still relevant:

> We should log in debug the diags, it will help us investigate some case where we wont be able to read the lockfile for some reason

We should not log something only when everything is fine, if there is some diags err we should log them in debug too.

**Review by eliecharra — CHANGES_REQUESTED** [review]
We have a looot of useless code in the lockfile read logic. Can you check and remove useless stuff, I think we can have simpler hcl decode logic

File: pkg/remote/terraform/provider.go (hunk 6 above)
Comment (sundowndev) [reply pkg/remote/terraform/provider.go:93]: I did add log messages in the LoadLocksFromFile func https://github.com/cloudskiff/driftctl/blob/4a455a1b9e25cf5ff0e91f3b042fcc6037b7931d/pkg/remote/terraform/lock/lockfile.go#L44-L55

Comment (eliecharra) [reply pkg/remote/terraform/provider.go:93]: My bad I missed this one 🙏🏻 Anyway, I think that we should not bind the lockfile pkg to a logging strategy. Since we log here when everything is fine, it makes more sense to log errors at the same place IMHO. Let's keep the lockfile logic return it's errors without any error handling logic within.

**Comment by eliecharra** [comment]
> We have a looot of useless code in the lockfile read logic. Can you check and remove useless stuff, I think we can have simpler hcl decode logic.

Maybe I wasn't enough explicit whith this sentence. In the actual code we are still depending of terraform code (for diags), we should not have any dependency.

We also do not need to do the whole lock file validation process while reading it, here we only want to read the version. The validation of the file is done by terraform.

I think we could simplify things a lot, I tried to write a minimal decoding code below as an example :

```go
func main() {
	parser := hclparse.NewParser()
	f, diags := parser.ParseHCLFile(".terraform.lock.hcl")
	if diags.HasErrors() {
		fmt.Errorf("%+v", diags)
		os.Exit(1)
	}

	schema := &hcl.BodySchema{
		Blocks: []hcl.BlockHeaderSchema{
			{
				Type:       "provider",
				LabelNames: []string{"address"},
			},
		},
	}

	content, diags := f.Body.Content(schema)
	if diags.HasErrors() {
		fmt.Errorf("%+v", diags)
		os.Exit(1)
	}
	for _, provider := range content.Blocks.OfType("provider") {
		providerAddr := provider.Labels[0]
		attrs, _ := provider.Body.JustAttributes()
		for _, attr := range attrs {
			val, diags := attr.Expr.Value(nil)
			if diags.HasErrors() {
				fmt.Errorf("%+v", diags)
				os.Exit(1)
			}
			if attr.Name == "version" {
				fmt.Printf("%s@%s\n", providerAddr, val.AsString())
			}
		}
	}
}
```

**Comment by eliecharra** [comment]
Even simpler way with this approach :

```go
type Provider struct {
	Address string `hcl:"address,label"`
	Version string `hcl:"version"`
}

type Lockfile struct {
	Providers []Provider `hcl:"provider,block"`
}

func main() {
	parser := hclparse.NewParser()
	f, diags := parser.ParseHCLFile(".terraform.lock.hcl")
	if diags.HasErrors() {
		fmt.Errorf("%+v", diags)
		os.Exit(1)
	}

	lock := &Lockfile{}
	_ = gohcl.DecodeBody(f.Body, nil, lock)

	fmt.Printf("%s : %s", lock.Providers[0].Address, lock.Providers[0].Version)
}
```

File: pkg/terraform/lock/lockfile.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	var lock Lockfile
+
+	parser := hclparse.NewParser()
+	f, diags := parser.ParseHCLFile(filename)
+	if diags.HasErrors() {
+		return &lock, diags
+	}
+
+	diags = gohcl.DecodeBody(f.Body, nil, &lock)
+	if diags != nil {
```
Comment (eliecharra) [inline pkg/terraform/lock/lockfile.go:53]: This is useless, you can return `return &lock, diags` directly

File: pkg/terraform/lock/lockfile_test.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+			filepath: "testdata/lockfile_invalid_type-3.hcl",
+			assert: func(locks *Lockfile, err error) {
+				provider := locks.Provider(&ProviderAddress{
+					Type:      "google-beta",
+					Namespace: "hashicorp",
+					Hostname:  "registry.terraform.io",
+				})
+
+				assert.Len(t, locks.Providers, 2)
+				assert.Equal(t, "2.71.0", provider.Version)
```
Comment (eliecharra) [inline pkg/terraform/lock/lockfile_test.go:108]: You should assert on `tt` and pass it as a parameter since you are running subtests `c.assert(tt, locks, err)`

```go
assert: func(tt *testing.T, locks *Lockfile, err error) {
```

File: pkg/terraform/lock/lockfile.go [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+func (p *ProviderAddress) String() string {
+	return fmt.Sprintf("%s/%s/%s", p.Hostname, p.Namespace, p.Type)
+}
+
+type Lockfile struct {
+	Providers []Provider `hcl:"provider,block"`
+}
+
+func (l *Lockfile) Provider(addr *ProviderAddress) *Provider {
```
Comment (eliecharra) [inline pkg/terraform/lock/lockfile.go:34]: We cant name everything provider 😅 `s/Provider/GetProviderByAddress/`

File: pkg/terraform/lock/lockfile.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+// ProviderAddress encapsulates a single provider type. In the future this will be
+// extended to include additional fields including Namespace and SourceHost
+type ProviderAddress struct {
+	Type      string
+	Namespace string
+	Hostname  svchost.Hostname
+}
+
+func (p *ProviderAddress) String() string {
+	return fmt.Sprintf("%s/%s/%s", p.Hostname, p.Namespace, p.Type)
```
Comment (eliecharra) [inline pkg/terraform/lock/lockfile.go:27]: `strings.Join(string[]{p.HostName, p.Namespace, p.Type}, "/")`

**Review by eliecharra — CHANGES_REQUESTED** [review]

File: go.mod [hunk 11]
```diff
@@ -7,7 +7,6 @@ require (
 	github.com/Azure/azure-sdk-for-go/sdk/azcore v0.19.0
 	github.com/Azure/azure-sdk-for-go/sdk/azidentity v0.11.0
 	github.com/Azure/azure-sdk-for-go/sdk/storage/armstorage v0.2.0
-	github.com/Azure/azure-sdk-for-go/sdk/to v0.1.4
```
[the remark this answers is not in the document]
Comment (sundowndev) [reply go.mod:10]: Are you sure? I feel like it's unused, running `go mod tidy` removed it

[the remark this answers is not in the document]
Comment (eliecharra) [reply go.mod:10]: https://github.com/cloudskiff/driftctl/blob/main/pkg/remote/azurerm/repository/storage_test.go#L192

[the remark this answers is not in the document]
Comment (sundowndev) [reply go.mod:10]: Ok the diff will disappear  after a rebase

[the remark this answers is not in the document]
Comment (sundowndev) [reply go.mod:10]: It's `github.com/Azure/azure-sdk-for-go/sdk/azcore/to` so we can safely remove it

**Review by eliecharra — APPROVED** [review]
That looks great 👍🏻 I'm gonna merge it when the doc PR will be ready
